File: csharp/1293.0_Shortest_Path_in_a_Grid_with_Obstacles_Elimination.cs
/*
Runtime: 142 ms, faster than 95.60% of C# online submissions for Shortest Path in a Grid with Obstacles Elimination.
Memory Usage: 37.7 MB, less than 97.80% of C# online submissions for Shortest Path in a Grid with Obstacles Elimination.
*/
public class Solution {
    static int [][] dirs = {new int[]{-1, 0}, new int[]{1, 0}, new int[]{0, -1}, new int[]{0, 1}};
    
    public int ShortestPath(int[][] grid, int k) {
        int m = grid.Length, n = grid[0].Length;    
        if (k >= m + n - 3) {
            return m + n - 2;
        }
        int[][][] distances = new int[m][][];
        for (int i = 0; i < m; i++) {
            distances[i] = new int[n][];
            for (int j = 0; j < n; j++) {
                distances[i][j] = new int[k + 1];
                Array.Fill(distances[i][j], int.MaxValue);
            }
        }
        distances[0][0][0] = 0;
        Queue<Tuple<int, int, int>> q = new Queue<Tuple<int, int, int>>();
        q.Enqueue(new Tuple<int, int, int>(0, 0, 0));
        while (q.Count > 0) {
            Tuple<int, int, int> state = q.Dequeue();
            int row = state.Item1, col = state.Item2, eliminations = state.Item3;
            int distance = distances[row][col][eliminations];
            if (row == m - 1 && col == n - 1) {
                return distance;
            }
            foreach (int[] dir in dirs) {
                int newRow = row + dir[0], newCol = col + dir[1];
                if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n) {
                    if (grid[newRow][newCol] == 0) {
                        if (distances[newRow][newCol][eliminations] == int.MaxValue) {
                            distances[newRow][newCol][eliminations] = distance + 1;
                            q.Enqueue(new Tuple<int, int, int>(newRow, newCol, eliminations));
                        }
                    } else {
                        if (eliminations < k && distances[newRow][newCol][eliminations + 1] == int.MaxValue) {
                            distances[newRow][newCol][eliminations + 1] = distance + 1;
                            q.Enqueue(new Tuple<int, int, int>(newRow, newCol, eliminations + 1));
                        }
                    }
                }
            }
        }
        return -1;
    }
}

File: problems/1293.0_Shortest_Path_in_a_Grid_with_Obstacles_Elimination.py
'''
BFS

Runtime: 67 ms, faster than 98.11% of Python3 online submissions for Shortest Path in a Grid with Obstacles Elimination.
Memory Usage: 15.1 MB, less than 81.26% of Python3 online submissions for Shortest Path in a Grid with Obstacles Elimination.
'''
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        if m == 1 and n == 1:
            return 0
        if k >= m + n - 3:
            return m + n - 2
        
        visited = set([0, 0, k])
        q = deque([(0, 0, k)])
        step = 0
        while q:
            step += 1
            for _ in range(len(q)):
                x, y, rest = q.popleft()
                for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                    nx, ny = x + dx, y + dy 
                    if 0 <= nx < m and 0 <= ny < n:
                        if grid[nx][ny] == 0 and (nx, ny, rest) not in visited:
                            if nx == m - 1 and ny == n - 1:
                                return step
                            q.append((nx, ny, rest))
                            visited.add((nx, ny, rest))
                        elif grid[nx][ny] == 1 and rest > 0 and (nx, ny, rest - 1) not in visited:
                            q.append((nx, ny, rest - 1))
                            visited.add((nx, ny, rest - 1))
        return -1

'''
Runtime: 127 ms, faster than 84.88% of Python3 online submissions for Shortest Path in a Grid with Obstacles Elimination.
Memory Usage: 15.1 MB, less than 81.26% of Python3 online submissions for Shortest Path in a Grid with Obstacles Elimination.
'''
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        if m == 1 and n == 1:
            return 0
        if k >= m + n - 3:
            return m + n - 2
        
        visited = set([0, 0, k])
        q = deque([(0, 0, k)])
        step = 0
        while q:
            for _ in range(len(q)):
                x, y, rest = q.popleft()
                if x == m - 1 and y == n - 1:
                    return step                
                for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                    nx, ny = x + dx, y + dy 
                    if 0 <= nx < m and 0 <= ny < n:
                        if grid[nx][ny] == 0 and (nx, ny, rest) not in visited:
                            q.append((nx, ny, rest))
                            visited.add((nx, ny, rest))
                        elif grid[nx][ny] == 1 and rest > 0 and (nx, ny, rest - 1) not in visited:
                            q.append((nx, ny, rest - 1))
                            visited.add((nx, ny, rest - 1))
            step += 1
        return -1


        